### PR TITLE
num: implement exact foundational value types (ADR-004, #22)

### PR DESCRIPTION
**Pull request by rustyeddy**
## Summary

Clean-room implementation of Trader's exact foundational numeric types per ADR-004:

- `num/internal/fixed`: private scaled-integer representation kernel — common 1e8 scale, checked int64 add/sub/neg/abs with explicit `MinInt64` rejection, widened 128-bit signed multiplication/division, all four ADR-004 rounding policies, exact decimal parse/format with no `float64` anywhere in the path.
- `num.Rate`, `num.Price`, `num.Quantity`, `num.Currency`, `num.Money` — semantic types built on the kernel. `Price`/`Quantity` are non-negative by construction; `Money` requires an explicit `Currency` and rejects cross-currency arithmetic; `Currency` validates structurally only (3–5 uppercase ASCII letters), with no embedded ISO-4217/crypto registry.
- Text and JSON encoding for all five types. `Price`/`Quantity`/`Rate` serialize as canonical decimal strings; `Money`'s JSON form is `{"amount":"...","currency":"..."}`, its text form is the distinct `"123.45 USD"` canonical string.
- Architecture enforcement tests: an AST-based scan rejecting any `float32`/`float64` under `num`, and a second, redundant check (on top of Go's `internal/` visibility) that no package outside `num` imports `num/internal/fixed`.

Each commit is coherent and leaves the test suite green; 8 commits total.

## Why

Authoritative prices, quantities, money, and rates must not depend on binary floating-point approximation, and must be exact across backtest, paper, and live paths. See `docs/arch/adr-004-exact-numeric-representation.org` and issue #22.

## Test plan

- [x] `go build ./...`
- [x] `go vet ./num/...` — clean
- [x] `gofmt -l num/` — clean
- [x] `go test -race ./num/...` — clean
- [x] `make check` — clean except two **pre-existing** failures in `test/internal/numericstudy` and `test/internal/quantitystudy` (unguarded generated-table markers referencing the frozen ADR-004 doc); confirmed identical on `main` prior to this branch, unrelated to #22, left untouched
- [x] Table-driven tests covering zero/positive/negative/boundary/`MinInt64`/overflow/underflow/div-by-zero/cross-currency/rounding-midpoint cases for every operation
- [x] 6 fuzz targets (`FuzzFormatParse`, `FuzzParseFormat`, `FuzzAddSub`, `FuzzMulDivScaled`, `FuzzPriceJSONRoundTrip`, `FuzzMoneyTextRoundTrip`), each run locally for 10–15s with no failures

## Documentation

- `num/doc.go` documents the `num` / `num/internal/fixed` boundary, the context-free-vs-context-dependent arithmetic split, and why `Percent`/`Ratio` are deferred.
- Per-type doc comments explain zero/negative-value policy and intentionally absent operations (e.g. no implicit currency conversion, no public `Rounding` selector yet).

## Intentionally deferred (not in scope for #22)

- `Percent`/`Ratio` semantic wrappers
- Public `num.Rounding` API (kernel supports all 4 policies; only half-even is exposed until a caller needs to select one)
- Database persistence adapters (scaled constructors are in place to support this later)
- Currency registry (ISO-4217/crypto) — explicitly kept out of `num` per architecture decision; owned by future instrument/venue packages
- CI matrix expansion for cross-platform determinism coverage

Issue: #22

### CODE REVIEW

**Review by rustyeddy — COMMENTED** [review]
This is a strong implementation overall: the package split is right, the widened arithmetic design is sensible, the semantic APIs are mostly coherent, and the test coverage is unusually thorough. I do not think it should merge yet because three issues conflict with ADR-004 or the agreed package boundary.

## 1. Public raw-scaled escape hatches violate the `num` boundary

`PriceFromScaled`, `QuantityFromScaled`, `RateFromScaled`, `MoneyFromScaled`, and each type's exported `Scaled()` method expose the 1e8 representation to every package importing `num`.

That conflicts with the architecture we explicitly settled:

> Code outside `num` expresses financial intent. Code inside `num/internal/fixed` implements exact representation mechanics.

It also conflicts with ADR-004's rule that raw scaled construction is reserved for trusted internal/persistence/test use and that public APIs do not expose raw scaled integers. Documentation cannot make an exported Go API narrowly scoped.

Please remove the exported `FromScaled` constructors and `Scaled()` accessors from this PR. Persistence adapters are already listed as deferred; when implemented, they should use a deliberately designed persistence boundary rather than making the representation part of the ordinary public `num` API.

## 2. Invalid `Currency` values serialize successfully

The zero value of `Currency` is documented as invalid, but `Currency.MarshalText` returns an empty byte slice with no error and `Currency.MarshalJSON` emits `""` with no error. That silently converts an invalid semantic value into a wire value.

Please make both marshal methods validate `c.IsValid()` and return `ErrInvalidCurrency` (or a dedicated missing-currency error if preferred). Parsing already rejects empty currency codes; writing should enforce the same invariant.

Also consider making `Money.String()` visibly represent invalid money rather than returning `"0 "`. `MarshalText` and `MarshalJSON` correctly reject invalid Money, so this is not the same severity, but `String()` currently looks almost like canonical output while missing its required currency.

## 3. Unknown rounding values silently become half-even

`roundMagnitude` uses a `default` branch as `RoundHalfEven`. Any invalid `Rounding` value therefore silently selects a valid policy. Although `Rounding` is internal today, this is exactly the kind of hidden policy fallback ADR-004 is trying to avoid.

Please validate the mode in `MulScaled`/`DivScaled` (or `mulDiv`) and return an explicit error for unsupported values. The `default` branch should not reinterpret an invalid enum as half-even.

## Non-blocking observations

- The `num/internal/fixed` package boundary is the right design.
- The exact parser correctly handles the full negative `int64` boundary without constructing an unrepresentable positive magnitude.
- The widened `math/bits` multiply/divide approach looks appropriate and avoids a float escape path.
- Keeping caller-selected rounding internal for now is the right scope.
- Deferring Percent/Ratio, persistence adapters, currency registries, and expanded CI is appropriate.
- The architecture import test duplicates Go's `internal` enforcement, but it documents intent and is harmless.

After the three changes above, I would be comfortable doing a final arithmetic/API pass and likely approving.

**Comment by rustyeddy** [comment]
Pushed a fix for all three review points (commit b6b14f0):

1. **Raw-scaled escape hatches removed.** `PriceFromScaled`, `QuantityFromScaled`, `RateFromScaled`, `MoneyFromScaled`, and every `Scaled()` accessor are gone from the public API. Tests that need `MinInt64`/`MaxInt64` boundary values now build them with unexported struct literals (test files are `package num`), so no public surface was added to keep testing convenient. `doc.go` now explains that a future persistence adapter should introduce its own narrowly scoped boundary rather than reuse a general-purpose one.

2. **Invalid `Currency` marshal.** `MarshalText`/`MarshalJSON` on `Currency` now check `IsValid()` and return `ErrInvalidCurrency` instead of silently writing `""`. Also tightened `Money.String()`: invalid `Money` (including the zero value) now renders as `"<invalid money>"` instead of `"0 "`, which looked like plausible canonical output.

3. **Silent rounding fallback.** `MulScaled`/`DivScaled` now validate `Rounding` at the entry point and return `ErrInvalidRounding` for anything outside the four named policies. `roundMagnitude`'s switch no longer treats "anything else" as half-even; its default case now panics, documented as unreachable given the entry-point check, rather than quietly picking a policy.

Also confirmed the CI `race` failure is pre-existing on `main` (checked out `main` directly and reproduced it — `test/internal/numericstudy`/`quantitystudy` expect table markers that `docs/arch/adr-004-exact-numeric-representation.org` no longer has since it was frozen). Unrelated to this PR; `num` itself is green under `-race`.

All `num` tests, fuzzers, `go vet`, and `gofmt` pass locally after the changes.

**Review by rustyeddy — COMMENTED** [review]
Verified the latest fix commit `b6b14f0` against the three prior review findings.

## Resolved

1. **Raw scaled API removed**
   - `PriceFromScaled`, `QuantityFromScaled`, `RateFromScaled`, and `MoneyFromScaled` are gone.
   - Exported `Scaled()` accessors are gone.
   - Representation-boundary tests now use package-private struct construction inside `package num`, which keeps the raw 1e8 representation out of the public API.

2. **Invalid Currency serialization now fails**
   - `Currency.MarshalText` and `Currency.MarshalJSON` validate `IsValid()` and return `ErrInvalidCurrency` for the zero/invalid value.
   - Regression tests cover both text and JSON behavior.
   - `Money.String()` now makes invalid money visible as `<invalid money>` instead of producing the misleading `0 ` form.

3. **Unknown rounding modes now fail explicitly**
   - `MulScaled` and `DivScaled` reject invalid modes with `ErrInvalidRounding`.
   - `RoundHalfEven` is now an explicit switch case rather than the default fallback.
   - Tests cover invalid rounding values.

I also rechecked the affected arithmetic and serialization paths and found no regression introduced by these changes.

### Minor non-blocking cleanup

The PR description still says database persistence is deferred because "scaled constructors are in place." That sentence is now stale because the public scaled constructors were correctly removed. Please update that wording before merge so the PR description matches the final API.

From the code-review standpoint, my prior blockers are resolved. I see no remaining blocker from this review.

File: num/internal/fixed/fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		}
+
+		// Scaling by exactly one is the identity, in both directions.
+		if b == Scale {
+			assert.Equalf(t, a, product, "MulScaled(%d, 1) is not the identity", a)
+		}
+
+		// Zero propagates rather than producing a spurious remainder.
+		if a == 0 || b == 0 {
+			assert.Zerof(t, product, "MulScaled(%d,%d) should be zero", a, b)
```
[the remark this answers is not in the document]
Comment (rustyeddy) [reply num/internal/fixed/fuzz_test.go:116]: This is incorrect.
